### PR TITLE
Terrain Upsampling

### PR DESCRIPTION
**Pull request by bagnell**
Upsamples terrain from created mesh, not the data received from the server.

- [x] Fix existing tests
- [x] Add tests for new functions in `TerrainEncoding`

### CODE REVIEW

**Comment by pjcozzi** [comment]
@kring do you want to look at this?

File: Source/Core/HeightmapTerrainData.js [hunk 1]
```diff
@@ -220,6 +226,8 @@ define([
                     6,
                     result.orientedBoundingBox,
                     TerrainEncoding.clone(result.encoding));
+            that._buffer = undefined;
```
Comment (pjcozzi) [inline Source/Core/HeightmapTerrainData.js]: Perhaps add a comment about why this is set to `undefined`.

**Comment by pjcozzi** [comment]
Did you test this in Firefox?  I am not seeing a memory difference between master and this branch in Windows Task Manager.  Perhaps the FF GC is just not as aggressive?

**Comment by pjcozzi** [comment]
No difference in IE 11 either.

I did not try Chrome.

**Comment by bagnell** [comment]
I saw only ~10% decrease in both IE and FF compared to the ~35% in Chrome.

**Comment by pjcozzi** [comment]
Do you think there are any other references to the arraybuffer?  Are there other views into the same buffer?

Did you try scenarios where, for example, lots of other memory allocations would create GC pressure so that the memory savings were noticeable, e.g., a fly-through instead of a single view?

Also, instead of setting the arraybuffer member to undefined, perhaps there is a way to restructure this so the arraybuffer is not part of the class.  I don't know enough about JavaScript engines to say if that would help.

**Comment by kring** [comment]
FWIW, when I originally wrote this code, `TerrainMesh` was transient.  We'd create it, upload it to the GPU, and throw it away.  My intention was that all upsampling, hit detection, etc. would happen against the terrain data (quantized or heightmap) rather than against the mesh.

I guess that changed when the terrain/camera intersection was added, because it was impractical to do intersections against the downloaded data quickly?  I can certainly imagine that being the case.

If so, it seems reasonable to go the other way: keep the mesh and throw away the downloaded data.  Keep in mind that this has an impact on loading, though.

Currently, for heightmap data, as soon as we download the data we can start upsampling it for child tiles.  This lets us drill down through the tree more quickly so that we can show the imagery more quickly (even if the terrain has low detail to start).  For quantized mesh data, the upsample is more expensive and involves a web worker, so we can't drill down as quickly.  But with this PR, we now need to run two processes in a web worker per quantized mesh tile, rather than one.  First we need to create a mesh from the data, then we need to upsample that mesh.  That means we'll drill down through the tree even more slowly than before.

Also, I'm concerned (but haven't convinced myself one way or the other yet) that this change might cause the load process to waste a bunch of CPU trying to upsample tiles that aren't ready to be upsampled yet, because the `upsample` method now returns undefined if `_mesh` isn't yet set.

Then again, this is all making me remember again how fundamentally flawed the terrain/imagery load code is, so maybe it won't be any worse at all (and certainly it will use less memory!). :)

**Comment by pjcozzi** [comment]
@kring all sounds OK to me.  Do you have any other feedback here?

We'll look into this:

> I'm concerned (but haven't convinced myself one way or the other yet) that this change might cause the load process to waste a bunch of CPU trying to upsample tiles that aren't ready to be upsampled yet, because the upsample method now returns undefined if _mesh isn't yet set.

**Comment by kring** [comment]
> Do you have any other feedback here?

No, it all looks good to me.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

**Comment by pjcozzi** [comment]
Update CHANGES.md to say this is specific to Chrome.

I tested the following view in master and this branch (wait for the imagery to fully load, then enable terrain and wait for it to fully load):

[http://localhost:8080/Apps/CesiumViewer/index.html?view=-79.36627737777864%2C38.8344057990329%2C973.3720887919472%2C319.5773668432903%2C-1.0443130952986466%2C359.87816292614366](http://localhost:8080/Apps/CesiumViewer/index.html?view=-79.36627737777864%2C38.8344057990329%2C973.3720887919472%2C319.5773668432903%2C-1.0443130952986466%2C359.87816292614366)

In IE 11 and Firefox 42, the tests were inconclusive using the Windows Task Manager.  If there's not an issue in our code, I suspect you are right that they have bugs releasing typed arrays.  Perhaps we could test this by allocating a fixed set of typed arrays up front and using them as a ring buffer.  That might even be a better strategy in general to avoid GC pressure at the cost of using a fixed chunk of memory and limiting the total number of tiles that can be in-flight.  Just an idea, it's fine to save this for a separate PR.

For Chrome, here's the memory difference I saw for the _tab_:
* `master`: 164 MB
* `upsample`: 124-132 MB

Given the high constants for everything in CPU, this is a really nice win.

**Comment by pjcozzi** [comment]
@bagnell now that Cesium 1.17 is out, let's wrap this up and aim to get it into master soon.

**Comment by bagnell** [comment]
@pjcozzi I think this is ready. I updated `CHANGES.md`. Is there any of comments I didn't address?

**Pull request merged**